docs/man/man3/controller.3
.TH "Controller" 3 "Wed Dec 7 2011" "Version 1" "upkeeper" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Controller \- 
.SS "Data Structures"

.in +1c
.ti -1c
.RI "struct \fB_buddy_info\fP"
.br
.RI "\fIinformation buddy returns. \fP"
.ti -1c
.RI "struct \fB_upk_controller_config\fP"
.br
.RI "\fIcontroller configuration. \fP"
.in -1c
.SS "Defines"

.in +1c
.ti -1c
.RI "#define \fBUPSERT_TEXT\fP(A)"
.br
.RI "\fIinsert or update a list option. \fP"
.ti -1c
.RI "#define \fBUPSERT_INT\fP(A, NMIN)"
.br
.ti -1c
.RI "#define \fBLIST_UPSERT_TEXT\fP(A, B, C)"
.br
.in -1c
.SS "Typedefs"

.in +1c
.ti -1c
.RI "typedef enum \fB_buddy_cmnd\fP \fBbuddy_cmnd_t\fP"
.br
.RI "\fIcontroller side of buddy protocol. \fP"
.ti -1c
.RI "typedef struct \fB_buddy_info\fP \fBbuddy_info_t\fP"
.br
.RI "\fIinformation buddy returns. \fP"
.ti -1c
.RI "typedef struct \fB_upk_controller_config\fP \fBupk_controller_config_t\fP"
.br
.RI "\fIcontroller configuration. \fP"
.in -1c
.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum \fB_buddy_cmnd\fP { \fBUPK_CTRL_NONE\fP =  0, \fBUPK_CTRL_ACK\fP =  1, \fBUPK_CTRL_SHUTDOWN\fP, \fBUPK_CTRL_STATUS_REQ\fP, \fBUPK_CTRL_ACTION_START\fP, \fBUPK_CTRL_ACTION_STOP\fP, \fBUPK_CTRL_ACTION_RELOAD\fP, \fBUPK_CTRL_ACTION_RUNONCE\fP, \fBUPK_CTRL_CUSTOM_ACTION_00\fP, \fBUPK_CTRL_CUSTOM_ACTION_01\fP, \fBUPK_CTRL_CUSTOM_ACTION_02\fP, \fBUPK_CTRL_CUSTOM_ACTION_03\fP, \fBUPK_CTRL_CUSTOM_ACTION_04\fP, \fBUPK_CTRL_CUSTOM_ACTION_05\fP, \fBUPK_CTRL_CUSTOM_ACTION_06\fP, \fBUPK_CTRL_CUSTOM_ACTION_07\fP, \fBUPK_CTRL_CUSTOM_ACTION_08\fP, \fBUPK_CTRL_CUSTOM_ACTION_09\fP, \fBUPK_CTRL_CUSTOM_ACTION_10\fP, \fBUPK_CTRL_CUSTOM_ACTION_11\fP, \fBUPK_CTRL_CUSTOM_ACTION_12\fP, \fBUPK_CTRL_CUSTOM_ACTION_13\fP, \fBUPK_CTRL_CUSTOM_ACTION_14\fP, \fBUPK_CTRL_CUSTOM_ACTION_15\fP, \fBUPK_CTRL_CUSTOM_ACTION_16\fP, \fBUPK_CTRL_CUSTOM_ACTION_17\fP, \fBUPK_CTRL_CUSTOM_ACTION_18\fP, \fBUPK_CTRL_CUSTOM_ACTION_19\fP, \fBUPK_CTRL_CUSTOM_ACTION_20\fP, \fBUPK_CTRL_CUSTOM_ACTION_21\fP, \fBUPK_CTRL_CUSTOM_ACTION_22\fP, \fBUPK_CTRL_CUSTOM_ACTION_23\fP, \fBUPK_CTRL_CUSTOM_ACTION_24\fP, \fBUPK_CTRL_CUSTOM_ACTION_25\fP, \fBUPK_CTRL_CUSTOM_ACTION_26\fP, \fBUPK_CTRL_CUSTOM_ACTION_27\fP, \fBUPK_CTRL_CUSTOM_ACTION_28\fP, \fBUPK_CTRL_CUSTOM_ACTION_29\fP, \fBUPK_CTRL_CUSTOM_ACTION_30\fP, \fBUPK_CTRL_CUSTOM_ACTION_31\fP, \fBUPK_CTRL_SIGNAL_01\fP, \fBUPK_CTRL_SIGNAL_02\fP, \fBUPK_CTRL_SIGNAL_03\fP, \fBUPK_CTRL_SIGNAL_04\fP, \fBUPK_CTRL_SIGNAL_05\fP, \fBUPK_CTRL_SIGNAL_06\fP, \fBUPK_CTRL_SIGNAL_07\fP, \fBUPK_CTRL_SIGNAL_08\fP, \fBUPK_CTRL_SIGNAL_09\fP, \fBUPK_CTRL_SIGNAL_10\fP, \fBUPK_CTRL_SIGNAL_11\fP, \fBUPK_CTRL_SIGNAL_12\fP, \fBUPK_CTRL_SIGNAL_13\fP, \fBUPK_CTRL_SIGNAL_14\fP, \fBUPK_CTRL_SIGNAL_15\fP, \fBUPK_CTRL_SIGNAL_16\fP, \fBUPK_CTRL_SIGNAL_17\fP, \fBUPK_CTRL_SIGNAL_18\fP, \fBUPK_CTRL_SIGNAL_19\fP, \fBUPK_CTRL_SIGNAL_20\fP, \fBUPK_CTRL_SIGNAL_21\fP, \fBUPK_CTRL_SIGNAL_22\fP, \fBUPK_CTRL_SIGNAL_23\fP, \fBUPK_CTRL_SIGNAL_24\fP, \fBUPK_CTRL_SIGNAL_25\fP, \fBUPK_CTRL_SIGNAL_26\fP, \fBUPK_CTRL_SIGNAL_27\fP, \fBUPK_CTRL_SIGNAL_28\fP, \fBUPK_CTRL_SIGNAL_29\fP, \fBUPK_CTRL_SIGNAL_30\fP, \fBUPK_CTRL_SIGNAL_31\fP, \fBUPK_CTRL_SIGNAL_32\fP, \fBUPK_CTRL_NONE\fP =  0, \fBUPK_CTRL_ACK\fP =  1, \fBUPK_CTRL_SHUTDOWN\fP, \fBUPK_CTRL_STATUS_REQ\fP, \fBUPK_CTRL_ACTION_START\fP, \fBUPK_CTRL_ACTION_STOP\fP, \fBUPK_CTRL_ACTION_RELOAD\fP, \fBUPK_CTRL_ACTION_RUNONCE\fP, \fBUPK_CTRL_CUSTOM_ACTION_00\fP, \fBUPK_CTRL_CUSTOM_ACTION_01\fP, \fBUPK_CTRL_CUSTOM_ACTION_02\fP, \fBUPK_CTRL_CUSTOM_ACTION_03\fP, \fBUPK_CTRL_CUSTOM_ACTION_04\fP, \fBUPK_CTRL_CUSTOM_ACTION_05\fP, \fBUPK_CTRL_CUSTOM_ACTION_06\fP, \fBUPK_CTRL_CUSTOM_ACTION_07\fP, \fBUPK_CTRL_CUSTOM_ACTION_08\fP, \fBUPK_CTRL_CUSTOM_ACTION_09\fP, \fBUPK_CTRL_CUSTOM_ACTION_10\fP, \fBUPK_CTRL_CUSTOM_ACTION_11\fP, \fBUPK_CTRL_CUSTOM_ACTION_12\fP, \fBUPK_CTRL_CUSTOM_ACTION_13\fP, \fBUPK_CTRL_CUSTOM_ACTION_14\fP, \fBUPK_CTRL_CUSTOM_ACTION_15\fP, \fBUPK_CTRL_CUSTOM_ACTION_16\fP, \fBUPK_CTRL_CUSTOM_ACTION_17\fP, \fBUPK_CTRL_CUSTOM_ACTION_18\fP, \fBUPK_CTRL_CUSTOM_ACTION_19\fP, \fBUPK_CTRL_CUSTOM_ACTION_20\fP, \fBUPK_CTRL_CUSTOM_ACTION_21\fP, \fBUPK_CTRL_CUSTOM_ACTION_22\fP, \fBUPK_CTRL_CUSTOM_ACTION_23\fP, \fBUPK_CTRL_CUSTOM_ACTION_24\fP, \fBUPK_CTRL_CUSTOM_ACTION_25\fP, \fBUPK_CTRL_CUSTOM_ACTION_26\fP, \fBUPK_CTRL_CUSTOM_ACTION_27\fP, \fBUPK_CTRL_CUSTOM_ACTION_28\fP, \fBUPK_CTRL_CUSTOM_ACTION_29\fP, \fBUPK_CTRL_CUSTOM_ACTION_30\fP, \fBUPK_CTRL_CUSTOM_ACTION_31\fP, \fBUPK_CTRL_SIGNAL_01\fP, \fBUPK_CTRL_SIGNAL_02\fP, \fBUPK_CTRL_SIGNAL_03\fP, \fBUPK_CTRL_SIGNAL_04\fP, \fBUPK_CTRL_SIGNAL_05\fP, \fBUPK_CTRL_SIGNAL_06\fP, \fBUPK_CTRL_SIGNAL_07\fP, \fBUPK_CTRL_SIGNAL_08\fP, \fBUPK_CTRL_SIGNAL_09\fP, \fBUPK_CTRL_SIGNAL_10\fP, \fBUPK_CTRL_SIGNAL_11\fP, \fBUPK_CTRL_SIGNAL_12\fP, \fBUPK_CTRL_SIGNAL_13\fP, \fBUPK_CTRL_SIGNAL_14\fP, \fBUPK_CTRL_SIGNAL_15\fP, \fBUPK_CTRL_SIGNAL_16\fP, \fBUPK_CTRL_SIGNAL_17\fP, \fBUPK_CTRL_SIGNAL_18\fP, \fBUPK_CTRL_SIGNAL_19\fP, \fBUPK_CTRL_SIGNAL_20\fP, \fBUPK_CTRL_SIGNAL_21\fP, \fBUPK_CTRL_SIGNAL_22\fP, \fBUPK_CTRL_SIGNAL_23\fP, \fBUPK_CTRL_SIGNAL_24\fP, \fBUPK_CTRL_SIGNAL_25\fP, \fBUPK_CTRL_SIGNAL_26\fP, \fBUPK_CTRL_SIGNAL_27\fP, \fBUPK_CTRL_SIGNAL_28\fP, \fBUPK_CTRL_SIGNAL_29\fP, \fBUPK_CTRL_SIGNAL_30\fP, \fBUPK_CTRL_SIGNAL_31\fP, \fBUPK_CTRL_SIGNAL_32\fP, \fBUPK_CTRL_NONE\fP =  0, \fBUPK_CTRL_ACK\fP =  1, \fBUPK_CTRL_SHUTDOWN\fP, \fBUPK_CTRL_STATUS_REQ\fP, \fBUPK_CTRL_ACTION_START\fP, \fBUPK_CTRL_ACTION_STOP\fP, \fBUPK_CTRL_ACTION_RELOAD\fP, \fBUPK_CTRL_ACTION_RUNONCE\fP, \fBUPK_CTRL_CUSTOM_ACTION_00\fP, \fBUPK_CTRL_CUSTOM_ACTION_01\fP, \fBUPK_CTRL_CUSTOM_ACTION_02\fP, \fBUPK_CTRL_CUSTOM_ACTION_03\fP, \fBUPK_CTRL_CUSTOM_ACTION_04\fP, \fBUPK_CTRL_CUSTOM_ACTION_05\fP, \fBUPK_CTRL_CUSTOM_ACTION_06\fP, \fBUPK_CTRL_CUSTOM_ACTION_07\fP, \fBUPK_CTRL_CUSTOM_ACTION_08\fP, \fBUPK_CTRL_CUSTOM_ACTION_09\fP, \fBUPK_CTRL_CUSTOM_ACTION_10\fP, \fBUPK_CTRL_CUSTOM_ACTION_11\fP, \fBUPK_CTRL_CUSTOM_ACTION_12\fP, \fBUPK_CTRL_CUSTOM_ACTION_13\fP, \fBUPK_CTRL_CUSTOM_ACTION_14\fP, \fBUPK_CTRL_CUSTOM_ACTION_15\fP, \fBUPK_CTRL_CUSTOM_ACTION_16\fP, \fBUPK_CTRL_CUSTOM_ACTION_17\fP, \fBUPK_CTRL_CUSTOM_ACTION_18\fP, \fBUPK_CTRL_CUSTOM_ACTION_19\fP, \fBUPK_CTRL_CUSTOM_ACTION_20\fP, \fBUPK_CTRL_CUSTOM_ACTION_21\fP, \fBUPK_CTRL_CUSTOM_ACTION_22\fP, \fBUPK_CTRL_CUSTOM_ACTION_23\fP, \fBUPK_CTRL_CUSTOM_ACTION_24\fP, \fBUPK_CTRL_CUSTOM_ACTION_25\fP, \fBUPK_CTRL_CUSTOM_ACTION_26\fP, \fBUPK_CTRL_CUSTOM_ACTION_27\fP, \fBUPK_CTRL_CUSTOM_ACTION_28\fP, \fBUPK_CTRL_CUSTOM_ACTION_29\fP, \fBUPK_CTRL_CUSTOM_ACTION_30\fP, \fBUPK_CTRL_CUSTOM_ACTION_31\fP, \fBUPK_CTRL_SIGNAL_01\fP, \fBUPK_CTRL_SIGNAL_02\fP, \fBUPK_CTRL_SIGNAL_03\fP, \fBUPK_CTRL_SIGNAL_04\fP, \fBUPK_CTRL_SIGNAL_05\fP, \fBUPK_CTRL_SIGNAL_06\fP, \fBUPK_CTRL_SIGNAL_07\fP, \fBUPK_CTRL_SIGNAL_08\fP, \fBUPK_CTRL_SIGNAL_09\fP, \fBUPK_CTRL_SIGNAL_10\fP, \fBUPK_CTRL_SIGNAL_11\fP, \fBUPK_CTRL_SIGNAL_12\fP, \fBUPK_CTRL_SIGNAL_13\fP, \fBUPK_CTRL_SIGNAL_14\fP, \fBUPK_CTRL_SIGNAL_15\fP, \fBUPK_CTRL_SIGNAL_16\fP, \fBUPK_CTRL_SIGNAL_17\fP, \fBUPK_CTRL_SIGNAL_18\fP, \fBUPK_CTRL_SIGNAL_19\fP, \fBUPK_CTRL_SIGNAL_20\fP, \fBUPK_CTRL_SIGNAL_21\fP, \fBUPK_CTRL_SIGNAL_22\fP, \fBUPK_CTRL_SIGNAL_23\fP, \fBUPK_CTRL_SIGNAL_24\fP, \fBUPK_CTRL_SIGNAL_25\fP, \fBUPK_CTRL_SIGNAL_26\fP, \fBUPK_CTRL_SIGNAL_27\fP, \fBUPK_CTRL_SIGNAL_28\fP, \fBUPK_CTRL_SIGNAL_29\fP, \fBUPK_CTRL_SIGNAL_30\fP, \fBUPK_CTRL_SIGNAL_31\fP, \fBUPK_CTRL_SIGNAL_32\fP }"
.br
.RI "\fIcontroller side of buddy protocol. \fP"
.ti -1c
.RI "enum \fBbuddy_runstate_t\fP { \fBBUDDY_UNKNOWN\fP =  0, \fBBUDDY_STOPPED\fP =  1, \fBBUDDY_RUNNING\fP, \fBBUDDY_RANONCE\fP, \fBBUDDY_UNKNOWN\fP =  0, \fBBUDDY_STOPPED\fP =  1, \fBBUDDY_RUNNING\fP, \fBBUDDY_RANONCE\fP, \fBBUDDY_UNKNOWN\fP =  0, \fBBUDDY_STOPPED\fP =  1, \fBBUDDY_RUNNING\fP, \fBBUDDY_RANONCE\fP }"
.br
.RI "\fIstates for buddy \fP"
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "int \fBupk_db_get_single_text\fP (\fBsqlite3\fP *dbh, char *value, const char *sql,...)"
.br
.RI "\fIlookup a table_id from the schema table. \fP"
.ti -1c
.RI "static int \fBupk_db_get_table_id\fP (\fBsqlite3\fP *dbh, const char *db, const char *table_name, char *table_id)"
.br
.RI "\fIlookup a table_id from the schema table. \fP"
.ti -1c
.RI "int \fBupk_db_get_audit_service_table\fP (\fBsqlite3\fP *dbh, char *table_id)"
.br
.RI "\fIlookup the table_id for upk_audit.service from the schema table. \fP"
.ti -1c
.RI "int \fBupk_db_get_cfg_services_table\fP (\fBsqlite3\fP *dbh, char *table_id)"
.br
.RI "\fIlookup the table_id for upk_cfg.services from the schema table. \fP"
.ti -1c
.RI "int \fBupk_db_get_cfg_svc_options_table\fP (\fBsqlite3\fP *dbh, char *table_id)"
.br
.RI "\fIlookup the table_id for upk_cfg.svc_optionsfrom the schema table. \fP"
.ti -1c
.RI "int \fBupk_db_get_cfg_svc_list_options_table\fP (\fBsqlite3\fP *dbh, char *table_id)"
.br
.RI "\fIlookup the table_id for upk_cfg.svc_list_options from the schema table. \fP"
.ti -1c
.RI "char * \fBupk_db_svc_uuid_lookup\fP (\fBsqlite3\fP *dbh, const char *svcid, const char *package)"
.br
.RI "\fIlookup a service <package>-<name> \fP"
.ti -1c
.RI "char * \fBupk_db_gen_uuid\fP (\fBsqlite3\fP *dbh)"
.br
.RI "\fIgenerate new unique service UUID. \fP"
.ti -1c
.RI "bool \fBupk_db_add_new_service\fP (\fBsqlite3\fP *dbh, \fBupk_svc_desc_t\fP *svc)"
.br
.RI "\fICreate new service definition. \fP"
.ti -1c
.RI "static int \fBupk_db_try_insert\fP (\fBsqlite3\fP *dbh, char *table_id, \fBupk_svc_desc_t\fP *svc, int type, char *optkey, void *optval, \fBsize_t\fP len, char *section, char *id)"
.br
.RI "\fItry to insert an option \fP"
.ti -1c
.RI "static int \fBupk_db_try_update\fP (\fBsqlite3\fP *dbh, char *table_id, \fBupk_svc_desc_t\fP *svc, int type, char *optkey, void *optval, \fBsize_t\fP len, char *section, char *id)"
.br
.RI "\fItry to update an option \fP"
.ti -1c
.RI "bool \fBupk_db_upsert\fP (\fBsqlite3\fP *dbh, char *table_id, \fBupk_svc_desc_t\fP *svc, int type, char *optkey, void *optval, \fBsize_t\fP len, char *section, char *id)"
.br
.RI "\fIinsert or update a option. \fP"
.ti -1c
.RI "int32_t \fBupk_db_upsert_options\fP (\fBsqlite3\fP *dbh, \fBupk_svc_desc_t\fP *svc)"
.br
.RI "\fIInsert options into database;. \fP"
.ti -1c
.RI "bool \fBupk_db_insert_cfg\fP (\fBsqlite3\fP *dbh, \fBupk_svc_desc_t\fP *svc)"
.br
.RI "\fIInsert a configuration into the database. \fP"
.in -1c
.SH "Define Documentation"
.PP 
.SS "#define LIST_UPSERT_TEXT(A, B, C)"\fBValue:\fP
.PP
.nf
do { \
            if(svc->A && svc->A->thisp && svc->A->thisp->B && strnlen(svc->A->thisp->B, UPK_MAX_STRING_LEN) > 0) { \
                n = strnlen(svc->A->thisp->B, UPK_MAX_STRING_LEN); \
                upk_db_upsert(dbh, table_id, svc, SQLITE_TEXT, stringify(B), svc->A->thisp->B, -1, stringify(A), C); \
                if(n > 0) { \
                    hash_buffer_len += n; \
                    hash_buffer = realloc(hash_buffer, hash_buffer_len); \
                    strncat(hash_buffer, svc->A->thisp->B, n); \
                } \
            } \
        } while(0)
.fi
.PP
Referenced by upk_db_upsert_options().
.SS "#define UPSERT_INT(A, NMIN)"\fBValue:\fP
.PP
.nf
do { \
            if(svc->A != NMIN) { \
                upk_db_upsert(dbh, table_id, svc, SQLITE_INTEGER, stringify(A), &svc->A, 0, NULL, NULL); \
                n = snprintf(tmp_buffer, UPK_MAX_STRING_LEN - 1, '%ld', (long int) svc->A); \
                if(n > 0) { \
                    hash_buffer_len += n;  \
                    hash_buffer = realloc(hash_buffer, hash_buffer_len); \
                    strncat(hash_buffer, tmp_buffer, n); \
                } \
            } \
        } while(0)
.fi
.PP
Referenced by upk_db_upsert_options().
.SS "#define UPSERT_TEXT(A)"\fBValue:\fP
.PP
.nf
do { \
            if(svc->A && strnlen(svc->A, UPK_MAX_STRING_LEN) > 0) { \
                n = strnlen(svc->A, UPK_MAX_STRING_LEN); \
                upk_db_upsert(dbh, table_id, svc, SQLITE_TEXT, stringify(A), svc->A, -1, NULL, NULL); \
                hash_buffer_len += n; \
                hash_buffer = realloc(hash_buffer, hash_buffer_len); \
                strncat(hash_buffer, svc->A, n); \
            } \
        } while(0)
.fi
.PP
insert or update a list option. ****************************************************************************************************************** 
.PP
\fBParameters:\fP
.RS 4
\fIdbh\fP The database handle. 
.br
\fIsvc\fP The service description to insert. 
.br
\fItype\fP The type to use for bind. 
.br
\fIoptkey\fP The name of the option to insert/update. 
.br
\fIoptval\fP The value to insert/update.
.RE
.PP
\fBReturns:\fP
.RS 4
true on success, false on failure. 
.RE
.PP

.PP
Referenced by upk_db_upsert_options().
.SH "Typedef Documentation"
.PP 
.SS "typedef enum \fB_buddy_cmnd\fP  \fBbuddy_cmnd_t\fP"
.PP
controller side of buddy protocol. This is what controller sends to buddy 
.SS "typedef struct \fB_buddy_info\fP \fBbuddy_info_t\fP"
.PP
information buddy returns. 
.SS "typedef struct \fB_upk_controller_config\fP  \fBupk_controller_config_t\fP"
.PP
controller configuration. 
.SH "Enumeration Type Documentation"
.PP 
.SS "enum \fB_buddy_cmnd\fP"
.PP
controller side of buddy protocol. This is what controller sends to buddy 
.PP
\fBEnumerator: \fP
.in +1c
.TP
\fB\fIUPK_CTRL_NONE \fP\fP
unset 
.TP
\fB\fIUPK_CTRL_ACK \fP\fP
controller received the last msg buddy sent it 
.TP
\fB\fIUPK_CTRL_SHUTDOWN \fP\fP
buddy shutdown 
.TP
\fB\fIUPK_CTRL_STATUS_REQ \fP\fP
report status to controller 
.TP
\fB\fIUPK_CTRL_ACTION_START \fP\fP
.TP
\fB\fIUPK_CTRL_ACTION_STOP \fP\fP
run stop script, which should kill the monitored process 
.TP
\fB\fIUPK_CTRL_ACTION_RELOAD \fP\fP
run reload script, which could, for example, send SIGHUP 
.TP
\fB\fIUPK_CTRL_ACTION_RUNONCE \fP\fP
run a monitored process in a restarter... only once... ??? 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_00 \fP\fP
run user-defined action 00 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_01 \fP\fP
run user-defined action 01 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_02 \fP\fP
run user-defined action 02 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_03 \fP\fP
run user-defined action 03 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_04 \fP\fP
run user-defined action 04 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_05 \fP\fP
run user-defined action 05 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_06 \fP\fP
run user-defined action 06 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_07 \fP\fP
run user-defined action 07 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_08 \fP\fP
run user-defined action 08 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_09 \fP\fP
run user-defined action 09 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_10 \fP\fP
run user-defined action 10 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_11 \fP\fP
run user-defined action 11 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_12 \fP\fP
run user-defined action 12 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_13 \fP\fP
run user-defined action 13 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_14 \fP\fP
run user-defined action 14 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_15 \fP\fP
run user-defined action 15 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_16 \fP\fP
run user-defined action 16 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_17 \fP\fP
run user-defined action 17 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_18 \fP\fP
run user-defined action 18 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_19 \fP\fP
run user-defined action 19 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_20 \fP\fP
run user-defined action 20 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_21 \fP\fP
run user-defined action 21 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_22 \fP\fP
run user-defined action 22 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_23 \fP\fP
run user-defined action 23 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_24 \fP\fP
run user-defined action 24 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_25 \fP\fP
run user-defined action 25 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_26 \fP\fP
run user-defined action 26 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_27 \fP\fP
run user-defined action 27 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_28 \fP\fP
run user-defined action 28 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_29 \fP\fP
run user-defined action 29 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_30 \fP\fP
run user-defined action 30 
.TP
\fB\fIUPK_CTRL_CUSTOM_ACTION_31 \fP\fP
run user-defined action 31 
.TP
\fB\fIUPK_CTRL_SIGNAL_01 \fP\fP

Changes

    (0.02) 2011/02/15
    (ms) Split database into main (control) and misc (listeners and 
         potential statistics), abstracted out filenames (reasonable 
         defaults, but settable via upk_db_file_xxx_set()), changed
         upk_db_init to use multi-db handle struct.

    (0.01) 2010/XX/XX
    (ms) Where it all began


controller/controller.c

/****************************************************************************
 * Copyright (c) 2011 Yahoo! Inc. All rights reserved. Licensed under the
 * Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License
 * at http://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable
 * law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 * See accompanying LICENSE file. 
 ************************************************************************** */

#include <upkeeper.h>
#include <sqlite3.h>
#include <sys/file.h>
#include "controller.h"
#include "ctrl_sql.h"
#include "ctrl_data.h"

sqlite3                *ctrl_dbh;
static char             flockpath[UPK_MAX_STRING_LEN] = "";
static int              flock_fd;

/** ******************************************************************************************************************
  @brief controller state initialization.
 ******************************************************************************************************************* */
void
upk_state_init(void)
{
    struct stat             st;
    int                     errno_tmp;
    int                     flock_fd = -1;

    errno = 0;
    upk_mkdir_p(upk_runtime_configuration.StateDir);
    errno = errno_tmp;
    if((stat(upk_runtime_configuration.StateDir, &st) != 0)
       || !S_ISDIR(st.st_mode)) {
        upk_fatal("State directory doesn't exist and could not be created: %s: %s\n", upk_runtime_configuration.StateDir, errno_tmp);
    }

    strncpy(flockpath, upk_runtime_configuration.StateDir, UPK_MAX_STRING_LEN - 1);
    strncat(flockpath, "/.lock", UPK_MAX_STRING_LEN - 1 - strnlen(upk_runtime_configuration.StateDir, UPK_MAX_STRING_LEN));

    errno = 0;
    flock_fd = open(flockpath, O_CREAT);
    if(errno)
        upk_fatal("Could not open %s: %s\n", flockpath, strerror(errno));

    fcntl(flock_fd, F_SETFD, FD_CLOEXEC);

    if(flock(flock_fd, LOCK_EX) != 0)
        upk_fatal("Another instance of controller is already running\n");

    return;
}

/** ******************************************************************************************************************
 ******************************************************************************************************************* */
static void
upk_db_path(char **pathbuf)
{
    strcat(*pathbuf, upk_runtime_configuration.StateDir);
    strcat(*pathbuf, "/db");

    errno = 0;
    if(mkdir(*pathbuf, 0700) != 0 && errno != EEXIST)
        upk_fatal("Cannot create database dir %s", *pathbuf);
}

/** ******************************************************************************************************************
 ******************************************************************************************************************* */
void
upk_ctrl_attach_db(sqlite3 * dbh, char *db_path, char *db_name)
{
    char                   *sqlerr;
    int                     rc = 0;
    char                    attach_str[UPK_MAX_PATH_LEN] = "";

    snprintf(attach_str, UPK_MAX_STRING_LEN, "ATTACH \"%s/%s.sqlite3\" AS \"%s\";", db_path, db_name, db_name);
    rc = sqlite3_exec(dbh, attach_str, NULL, NULL, &sqlerr);
    if(rc != SQLITE_OK) {
        upk_fatal("failed to execute: %d: %s\n", rc, sqlerr);
        sqlite3_free(sqlerr);
    }
}

/** ******************************************************************************************************************
 ******************************************************************************************************************* */
sqlite3                *
upk_ctrl_init_db(char *db_path)
{
    sqlite3                *dbh = NULL;
    int                     rc = 0;
    char                   *db_pathp = NULL;
    char                   *sqlerr = NULL;

    // char table_id[UPK_MAX_STRING_LEN] = "";

    db_pathp = db_path + strnlen(db_path, UPK_MAX_PATH_LEN);

    strcpy(db_pathp, "/upkeeper.sqlite3");
    if((rc = sqlite3_open(db_path, &dbh)) != 0) {
        upk_fatal("Cannot open database: %d\n", rc);
    }

    *db_pathp = '\0';
    upk_ctrl_attach_db(dbh, db_path, "upk_cfg");
    upk_ctrl_attach_db(dbh, db_path, "upk_audit");

    rc = sqlite3_exec(dbh, (char *) ctrl_create_audit_sql, NULL, NULL, &sqlerr);
    if(rc != SQLITE_OK) {
        upk_fatal("failed to execute: %d: %s\n", rc, sqlerr);
        sqlite3_free(sqlerr);
    }
    rc = sqlite3_exec(dbh, (char *) ctrl_create_cfg_sql, NULL, NULL, &sqlerr);
    if(rc != SQLITE_OK) {
        upk_fatal("failed to execute: %d: %s\n", rc, sqlerr);
        sqlite3_free(sqlerr);
    }

    return dbh;
}

/** ******************************************************************************************************************
 ******************************************************************************************************************* */
void
upk_ctrl_exit(void)
{
    if(ctrl_dbh)
        sqlite3_close(ctrl_dbh);
    upk_ctrl_free_config();
    close(flock_fd);
    unlink(flockpath);
}


/** ******************************************************************************************************************
 ******************************************************************************************************************* */
int
upk_ctrl_init(void)
{
    char                    ctrl_db_path[UPK_MAX_PATH_LEN] = "", *p;

    p = ctrl_db_path;

    upk_ctrl_load_config();
    upk_state_init();
    upk_db_path(&p);

    atexit(upk_ctrl_exit);
    ctrl_dbh = upk_ctrl_init_db(ctrl_db_path);
    upk_load_runtime_services();

    UPKLIST_FOREACH(upk_file_configuration.svclist) {
        upk_db_insert_cfg(ctrl_dbh, upk_file_configuration.svclist->thisp);
    }
    return 0;
}
